cloudstack_plugin/volume.py

```
@@ -0,0 +1,134 @@
+                                      "volume is needed")
+
+        if 'size' in volume:
+            volume_size = volume['size']
+        else:
+            raise NonRecoverableError("To create a volume, the size of the "
+                                      "volume is needed")
+
+        volume = cloud_driver.create_volume(name=volume_name,
+                                            size=volume_size)
+
+        ctx.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY] = volume.id
+        ctx.instance.runtime_properties[CLOUDSTACK_TYPE_PROPERTY] = \
+            VOLUME_CLOUDSTACK_TYPE
+
+    elif ctx.node.properties[USE_EXTERNAL_RESOURCE_PROPERTY] is True:
+
+        if 'resource_id' in ctx.node.properties:
+            resource_id = ctx.node.properties['resource_id']
+        else:
+            raise NonRecoverableError("Resource_id is not supplied")
+
+        volume = get_volume_by_id(cloud_driver, resource_id)
+
+        ctx.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY] = volume.id
+        ctx.instance.runtime_properties[CLOUDSTACK_NAME_PROPERTY] = volume.name
+        ctx.instance.runtime_properties[CLOUDSTACK_TYPE_PROPERTY] = \
+            VOLUME_CLOUDSTACK_TYPE
+
+        return
+
+
+@operation
+def delete(**kwargs):
+    """ Delete a volume
+    """
+
+    cloud_driver = get_cloud_driver(ctx)
+
+    volume_id = ctx.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+    volume = get_volume_by_id(cloud_driver, volume_id)
+
+    if volume is None:
+        raise NonRecoverableError('Volume with id {0} not found'
+                                  .format(volume_id))
+    try:
+        if not ctx.node.properties[USE_EXTERNAL_RESOURCE_PROPERTY]:
+            ctx.logger.info('Trying to destroy volume {0}'.format(volume))
+            cloud_driver.destroy_volume(volume=volume)
+        else:
+            ctx.logger.info('Volume {0} does not need to be destroyed'.format(
+                volume))
+
+    except Exception as e:
```

no need to catch an exception here. its an actual error...let it be raised.